### PR TITLE
Refactor for explicit dialect registration

### PR DESCRIPTION
**Pull request by marbre**
Progress on #2958.

### CODE REVIEW

File: iree/tools/init_compiler_modules.h [hunk 1]
```diff
@@ -22,16 +22,22 @@
 namespace mlir {
 namespace iree_compiler {
 
+// Add all the IREE compiler module dialects to the provided registry.
+inline void registerIreeCompilerModuleDialects(DialectRegistry &registry) {
+  // clang-format off
```
Comment (GMNGeoffrey) [inline iree/tools/init_compiler_modules.h:27]: Why `clang-format off`?

File: iree/tools/init_compiler_modules.h [hunk 2]
```diff
@@ -22,16 +22,22 @@
...
-  static bool init_once = []() {
-    registerDialect<IREE::Check::CheckDialect>();
-    registerDialect<IREE::Strings::StringsDialect>();
-    registerDialect<IREE::TensorList::TensorListDialect>();
-    return true;
-  }();
-  (void)init_once;
+  static bool initOnce =
+      ([]() { registerIreeCompilerModuleDialects(getGlobalDialectRegistry()); }(),
+       true);
```
Comment (GMNGeoffrey) [inline iree/tools/init_compiler_modules.h:39]: I found the version with an explicit return more readable :-)

**Comment by GMNGeoffrey** [comment]
Sorry for the delay here. When I first looked it still said "draft" so I wasn't sure if it was ready

File: iree/tools/init_compiler_modules.h (hunk 1 above)
Comment (marbre) [reply iree/tools/init_compiler_modules.h:27]: This is taken over from upstreamÖ https://github.com/llvm/llvm-project/blob/05777ab941063192b9ccb1775358a83a2700ccc1/mlir/include/mlir/InitAllDialects.h#L41

File: iree/tools/init_compiler_modules.h (hunk 2 above)
Comment (marbre) [reply iree/tools/init_compiler_modules.h:39]: This is inspired by upstream as well https://github.com/llvm/llvm-project/blob/05777ab941063192b9ccb1775358a83a2700ccc1/mlir/include/mlir/InitAllDialects.h#L67
But we can go for an other style in IREE :-)

**Comment by marbre** [comment]
> Sorry for the delay here. When I first looked it still said "draft" so I wasn't sure if it was ready

Yeah, at it probably was at that time :) I changed from draft to review and assigned reviewers afterwards. Prior I just used the draft mode to trigger the build pipeline.

File: iree/tools/init_compiler_modules.h (hunk 2 above)
Comment (GMNGeoffrey) [reply iree/tools/init_compiler_modules.h:39]: Well it will probably all go away soon.

**Pull request merged**